File: monitoring/logging/logger.py
import logging
import os
from logging.handlers import TimedRotatingFileHandler

# Define the directory for logs
log_directory = "/var/log/anomaly_detection"
os.makedirs(log_directory, exist_ok=True)

# Logger configuration
logger = logging.getLogger("AnomalyDetectionLogger")
logger.setLevel(logging.DEBUG)

# Formatter for logs
formatter = logging.Formatter('%(asctime)s - %(name)s - %(levelname)s - %(message)s')

# File handler for rotating logs daily and keeping 7 backups
file_handler = TimedRotatingFileHandler(
    os.path.join(log_directory, "anomaly_detection.log"),
    when="midnight",
    interval=1,
    backupCount=7
)
file_handler.setFormatter(formatter)
file_handler.setLevel(logging.INFO)

# Stream handler for console output
console_handler = logging.StreamHandler()
console_handler.setFormatter(formatter)
console_handler.setLevel(logging.DEBUG)

# Add handlers to the logger
logger.addHandler(file_handler)
logger.addHandler(console_handler)

# Usage
def log_info(message):
    logger.info(message)

def log_error(message):
    logger.error(message)

def log_debug(message):
    logger.debug(message)

def log_warning(message):
    logger.warning(message)

log_info("Logger initialized successfully")
log_error("An error occurred in the anomaly detection model")

File: monitoring/logging/logger.R
library(futile.logger)

# Set up logger configuration
flog.appender(appender.file("log_file.log"))  # Log output to a file
flog.threshold(DEBUG)  # Set the threshold for logging (DEBUG, INFO, WARN, ERROR, FATAL)

# Custom log formatter (to customize log format)
formatter <- function(level, msg) {
  sprintf("[%s] %s: %s", format(Sys.time(), "%Y-%m-%d %H:%M:%S"), level, msg)
}
flog.layout(formatter)

# Log different levels of messages
flog.debug("Debug message: Starting anomaly detection process")
flog.info("Info message: Model prediction started")
flog.warn("Warning message: Anomaly detected in time-series data")
flog.error("Error message: Failed to preprocess data due to missing values")
flog.fatal("Fatal message: Model evaluation failed due to memory overflow")

# Function with embedded logging
run_anomaly_detection <- function(data) {
  flog.info("Processing data for anomaly detection")
  
  tryCatch({
    # Processing
    flog.debug("Running anomaly detection model...")
    
    # Model prediction logic
    anomalies <- detect_anomalies(data)  # This function should be defined elsewhere
    
    if (length(anomalies) == 0) {
      flog.info("No anomalies detected in the data.")
    } else {
      flog.warn(paste("Anomalies detected: ", length(anomalies)))
    }
    
  }, error = function(e) {
    flog.error(paste("Error in anomaly detection:", e$message))
  })
  
  flog.info("Anomaly detection process completed.")
}

# Function to clean up logs (e.g., rotation or periodic cleanup)
clean_logs <- function(log_file, max_size_mb = 10) {
  if (file.exists(log_file)) {
    log_size <- file.info(log_file)$size / (1024 * 1024)  # Convert to MB
    if (log_size > max_size_mb) {
      flog.info(paste("Log file exceeded max size. Rotating log..."))
      file.rename(log_file, paste0(log_file, "_backup_", format(Sys.time(), "%Y%m%d%H%M%S")))
      flog.appender(appender.file(log_file))  # Reset the log appender
    }
  }
}

# Usage
run_anomaly_detection(data_frame_here)
clean_logs("log_file.log")